client_request_metrics_test.py

```diff
@@ -0,0 +1,784 @@
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from functools import partial
+from itertools import repeat
+import pytest
+import time
+
+from dtest import Tester
+from tools.jmxutils import (JolokiaAgent, make_mbean)
+
+from cassandra import (ReadFailure, ReadTimeout, Unavailable,
+                       WriteFailure, WriteTimeout,
+                       ConsistencyLevel as CL)
+from cassandra.concurrent import execute_concurrent_with_args
+from cassandra.marshal import int32_pack
+from cassandra.policies import FallthroughRetryPolicy
+from cassandra.query import SimpleStatement
+
+since = pytest.mark.since
+
+jmx = None
+
+KEYSPACE = 'ks'
+FAIL_WRITE_KEYSPACE = 'fail_keyspace'
+VIEW_KEYSPACE = 'view_keyspace'
+TABLE = 't'
+VIEW = 'mv'
+TOMBSTONE_FAILURE_THRESHOLD = 20
+TOMBSTONE_FAIL_KEY = 10000001
+
+
+class NoException(BaseException):
+    pass
+
+
+@since('4.0')
+class TestClientRequestMetrics(Tester):
+
+    @pytest.fixture(autouse=True)
+    def fixture_add_additional_log_patterns(self, fixture_dtest_setup):
+        fixture_dtest_setup.ignore_log_patterns = (
+            'Testing write failures',  # The error to simulate a write failure
+            'ERROR WRITE_FAILURE',  # Logged in DEBUG mode for write failures
+            f"Scanned over {TOMBSTONE_FAILURE_THRESHOLD+1} tombstones during query"  # Caused by the read failure tests
+        )
+
+    def setup_once(self):
+        cluster = self.cluster
+        cluster.set_configuration_options({'read_request_timeout_in_ms': 3000,
+                                           'write_request_timeout_in_ms': 3000,
+                                           'phi_convict_threshold': 12,
+                                           'tombstone_failure_threshold': TOMBSTONE_FAILURE_THRESHOLD,
+                                           'enable_materialized_views': 'true'})
+        cluster.populate(2, debug=True)
+        self.start_cluster()
+        node1 = cluster.nodelist()[0]
+
+        global jmx
+        jmx = JolokiaAgent(node1)
+        jmx.start()
+
+        s = self.session = self.patient_exclusive_cql_connection(node1, retry_policy=FallthroughRetryPolicy(), request_timeout=30)
+        for k in [KEYSPACE, FAIL_WRITE_KEYSPACE]:
+            s.execute(f"CREATE KEYSPACE {k} WITH replication = {{'class': 'SimpleStrategy', 'replication_factor': '2'}}")
```

Comment:
Nit: we could use the more usual `create_ks` function:
```suggestion
            create_ks(s, k, 2)
```